ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/backwardsync/BackwardChain.java

```diff
@@ -86,22 +120,43 @@ public synchronized List<BlockHeader> getFirstNAncestorHeaders(final int size) {
   }
 
   public synchronized void prependAncestorsHeader(final BlockHeader blockHeader) {
-    if (firstStoredAncestor.isEmpty()) {
-      firstStoredAncestor = Optional.of(blockHeader);
-      lastStoredPivot = Optional.of(blockHeader);
+    prependAncestorsHeader(blockHeader, false);
+  }
+
+  public synchronized void prependAncestorsHeader(
+      final BlockHeader blockHeader, final boolean alreadyStored) {
+    if (!alreadyStored) {
       headers.put(blockHeader.getHash(), blockHeader);
-      return;
     }
-    final BlockHeader firstHeader = firstStoredAncestor.get();
-    headers.put(blockHeader.getHash(), blockHeader);
-    chainStorage.put(blockHeader.getHash(), firstHeader.getHash());
-    firstStoredAncestor = Optional.of(blockHeader);
-    LOG.atDebug()
-        .setMessage("Added header {} to backward chain led by pivot {} on height {}")
-        .addArgument(blockHeader::toLogString)
-        .addArgument(() -> lastStoredPivot.orElseThrow().toLogString())
-        .addArgument(firstHeader::getNumber)
-        .log();
+
+    if (firstStoredAncestor.isEmpty()) {
+      updateLastStorePivot(Optional.of(blockHeader));
```

Comment:
Why don't we add the blockHeader to the chainStorage here as well?

Does firstStoredAncestor.isEmpty represent the first BWS block that we've received?

Reply:
> Why don't we add the blockHeader to the chainStorage here as well?

`chainStorage` is the sequence of the blocks to import, each entry is blockHash -> nextBlockHash, so you know what block to import next, while the block headers are saved in another table.

> Does firstStoredAncestor.isEmpty represent the first BWS block that we've received?

it is empty at the beginning of the bws, and then it is updated when going backward/forward to point to the current block, and this is one of the variables that was not stored, but is required for to resume the session